src/include/storage/index/design_doc.md

```diff
@@ -0,0 +1,69 @@
+
+### Blocking manner
+
+A naive method to support building an index is locking the whole table to block any modifications on that table, building the index (including insertion a record into system catalogs) and releasing the lock of the table after building the index. Those operations can be done within a single transaction. The deletion of the index is a single transaction of deleting the corresponding record in system catalogs.
+
+The challenging part is the semantics of the lock. The semantics should be
+* During creation of an index, all modifications including insertion, deletion and update on the table should be blocked.
+* When the system is not creating the index, all modifications including insertion, deletion and update on the table should not block one another (at least snapshot isolation).
+* At any time, scan of the table should not be blocked.
+
+After that, a shared/exclusive lock with the granularity of a table corresponds that semantics. When creation of an index starts, the transaction should first acquire the exclusive lock on that table. When the transaction completes creating the index, it releases the exclusive lock. When any modifications to the table starts, it should first acquire the shared lock on that table. When the modification completes, it releases the shared lock. All scanning operations will not come to the lock manager and read the table directly.
+
+This design is simple and straightforward. It is also easy to implement correctly. However, it may create a huge overhead when building an index on a extremely large table.
+
+### Non-Blocking manner
+
+The idea of creating index without blocking modifications on the table comes from [Postgres](https://www.postgresql.org/docs/11/sql-createindex.html). We use the similar idea here with small modifications . The creation of an index will be completed in two transactions:
+* The first transaction adds a record corresponding to the index into the system catalogs and sets the status of index to be ```READY```(allow insert/update but does not allow deletes nor queries)
+* The second transaction takes the scan of the whole target table and builds the index by adding all entries of its snapshot into the index. After that the transaction sets the status of the index in system catalogs to be ```VALID``` and commits.
+
+Note that before the start of the second transaction, the building process should wait for all running transactions to terminate, whose timestamps are smaller than the commit timestamp of the first transaction in the building process (the transaction adding a new entry on index into catalog).
+
```

Comment:
Why do you need to separate the two phases into two transactions? An alternative way to do it is to start one transaction, which first inserts into the catalog, and then builds the index on its visible version. What is the problem of this simpler method?

Reply:
Cuz during this period of time, other txns might also be working on the same table, but they cannot see that invisible index catalog entry. The index's content has to catch up with the most recent txns, but it is never going to happen if there's only one txn for building indexes.